packages/angular/build/src/utils/index-file/auto-csp_spec.ts

```diff
@@ -0,0 +1,170 @@
+    expect(csps[0]).toContain(hashTextContent("console.log('foo');"));
+  });
+
+  it('should rewrite a single source script', async () => {
+    const result = await autoCsp(`
+      <html>
+        <head>
+        </head>
+        <body>
+          <script src="./main.js"></script>
+          <div>Some text </div>
+        </body>
+      </html>
+    `);
+
+    const csps = getCsps(result);
+    expect(csps.length).toBe(1);
+    expect(csps[0]).toMatch(ONE_HASH_CSP);
+    expect(result).toContain(`var scripts = [['./main.js', undefined, false, false]];`);
+  });
+
+  it('should rewrite a single source script in place', async () => {
+    const result = await autoCsp(`
+      <html>
+        <head>
+        </head>
+        <body>
+          <div>Some text</div>
+          <script src="./main.js"></script>
+        </body>
+      </html>
+    `);
+
+    const csps = getCsps(result);
+    expect(csps.length).toBe(1);
+    expect(csps[0]).toMatch(ONE_HASH_CSP);
+    // Our loader script appears after the HTML text content.
+    expect(result).toMatch(
+      /Some text<\/div>\s*<script>\s*var scripts = \[\['.\/main.js', undefined, false, false\]\];/,
+    );
+  });
+
+  it('should rewrite a multiple source scripts with attributes', async () => {
+    const result = await autoCsp(`
+      <html>
+        <head>
+        </head>
+        <body>
+          <script src="./main1.js"></script>
+          <script async src="./main2.js"></script>
+          <script type="module" async defer src="./main3.js"></script>
+          <script type="application/not-javascript" src="./main4.js"></script>
+          <div>Some text </div>
+        </body>
+      </html>
+    `);
+
+    const csps = getCsps(result);
+    expect(csps.length).toBe(1);
+    expect(csps[0]).toMatch(ONE_HASH_CSP);
+    expect(result).toContain(
+      `var scripts = [['./main1.js', undefined, false, false],['./main2.js', undefined, true, false],['./main3.js', 'module', true, true]];`,
+    );
+    // Only one loader script is created.
+    expect(Array.from(result.matchAll(/\<script\>/g)).length).toEqual(1);
+  });
+
+  it('should rewrite source scripts with weird URLs', async () => {
+    const result = await autoCsp(`
+      <html>
+        <head>
+        </head>
+        <body>
+          <script src="/foo&amp;bar"></script>
+          <script src="/one'two\\'three\\\\'four\\\\\\'five"></script>
+          <script src="/one&two&amp;three&amp;amp;four"></script>
+          <script src="./</script>"></script>
+          <div>Some text </div>
+        </body>
+      </html>
+    `);
+
+    const csps = getCsps(result);
+    expect(csps.length).toBe(1);
+    expect(csps[0]).toMatch(ONE_HASH_CSP);
+    // &amp; encodes correctly
+    expect(result).toContain(`'/foo&bar'`);
+    // Impossible to escape a string and create invalid loader JS with a '
+    // (Quotes and backslashes work)
+    expect(result).toContain(`'/one\\'two%5C\\'three%5C%5C\\'four%5C%5C%5C\\'five'`);
```

Comment:
Question: Why do we want to `encodeURI` and convert `\` to `%5C`? If I do `<script src="/foo\bar">` and `<script src="/foo%5Cbar">` the browser seems to request two different URLs. I feel like we'd want to just propagate the `\` character?

Reply:
That is... certainly unfortunate, I am perplexed why this is the case? I was under the impression that `encodeURI` was spec-compliant to only encode characters that would be treated the same.

What are our options here? If we move away from encodeURI, I am not sure what clean options we have that are along the lines of "escape everything... except for %5C".

Reply:
I think we are going to live with backslashes being `%5C` for now, as it is RFC 3986 compliant.

Reply:
I'm ok with `%5C`, I'm inclined to go with that for now and revisit if we discover any problems.

Reply:
Keeping the escapaing and test cases as-is for now! Thanks!